[RDY] build/windows: Simplify build requirements on Windows

b-r-o-c-k opened the pull request:
After attempting to build on a clean install of Windows 10, I realized Perl is also a required dependency for generating Libvterm include files. I was able to rewrite most of the Perl script in CMake, although I did encounter a problem with converting characters to code points. It might be possible to do it in CMake by reading the file as a hex string and then implementing a simple UTF-8 decoder, but for simplicity I decided to use a PowerShell command instead.

justinmk commented on third-party/cmake/Libvterm-tbl2inc_c.cmake:
```diff
@@ -0,0 +1,58 @@
+  message("char: ${char}")
+  if(char)
+    # CMake doesn't seem to have an 'ord' function,
+    # so this PowerShell command is used to get the code point
```
Perhaps the entire script should be powershell, instead of cmake + powershell.

b-r-o-c-k committed:
build/windows: Ignore whitespace when applying patches

The --ignore-whitespace argument was added when applying patches to
prevent failures when patched files have different line endings.

b-r-o-c-k committed:
build/windows: Add CMakeLists.txt for Libuv on Windows

Adding a CMakeLists.txt for Libuv removes the need for Python
when building on Windows.

b-r-o-c-k committed:
build/windows: Add CMake script to generate headers for Libvterm

On Windows the CMake script will replace the Perl script used by
Libvterm to generate headers.

b-r-o-c-k committed:
build/windows: Add CMakeLists.txt for gperf

b-r-o-c-k committed:
build: Fix CMake target dependency problem

nvim was being ran before its runtime dependencies were copied.

b-r-o-c-k commented:
I didn't want PowerShell to be a requirement either, so I rewrote the entire Perl script in CMake. It's more complicated, but I think it's best not to have any build requirements other than a compiler and CMake.

justinmk commented on third-party/cmake/LibvtermCMakeLists.txt:
```diff
@@ -18,7 +18,7 @@ foreach(file ${TBL_FILES})
   add_custom_command(OUTPUT
     COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_BINARY_DIR}/encoding/
-    COMMAND ${PERL_EXECUTABLE} -CSD ${CMAKE_SOURCE_DIR}/tbl2inc_c.pl ${file} > ${CMAKE_BINARY_DIR}/${tname}
+    COMMAND ${CMAKE_COMMAND} -P ${CMAKE_SOURCE_DIR}/tbl2inc_c.cmake ${file} ${CMAKE_BINARY_DIR}/${tname}
```
If it's not much trouble, would be nice in the `perl_FOUND` case to compare the output of `tbl2inc_c.pl` to that of `tbl2inc_c.cmake`. As a sanity check. Then we know there's no bugs in `tbl2inc_c.cmake`. (Amazing effort, by the way!)

b-r-o-c-k commented on third-party/cmake/Libvterm-tbl2inc_c.cmake:
```diff
@@ -1,6 +1,6 @@
 cmake_minimum_required(VERSION 2.8.7)
 
-set(HEX_ALPHABET "0123456789ABCDEF")
+set(HEX_ALPHABET "0123456789abcdef")
```
I had to make the hex numbers lowercase and pad them so the output would be completely identical.

ZyX-I commented:
@b-r-o-c-k We have lua as a compile-time dependency, should be possible to use that even though it may appear it is installed in third-party. Though lua neither has UTF-8 decoder in a standart library.

ZyX-I commented on third-party/cmake/Libvterm-tbl2inc_c.cmake:
```diff
@@ -0,0 +1,163 @@
+    if(_resLen LESS ${padding})
+      math(EXPR _neededPadding "${padding} - ${_resLen}")
+      foreach(i RANGE 1 ${_neededPadding})
+	set(_res "0${_res}")
```
I see this unindented for some reason.

b-r-o-c-k replied:
Sorry, I accidentally inserted tabs in some places because I wasn't using `expandtab`. It's fixed now.

justinmk commented on third-party/cmake/BuildLibuv.cmake:
```diff
@@ -86,12 +81,19 @@ elseif(WIN32 AND MSVC)
   else()
     set(TARGET_CONFIG release)
   endif()
```
Is `TARGET_CONFIG` needed anymore?

b-r-o-c-k replied:
No, and neither is `TARGET_ARCH` so I removed them.

b-r-o-c-k commented on third-party/cmake/BuildGperf.cmake:
```diff
@@ -36,15 +36,28 @@ function(BuildGperf)
     INSTALL_COMMAND "${_gperf_INSTALL_COMMAND}")
 endfunction()
 
-set(GPERF_BUILDARGS CC=${HOSTDEPS_C_COMPILER} LD=${HOSTDEPS_C_COMPILER})
```
I removed this variable because it appears to be unused.

b-r-o-c-k commented:
@ZyX-I Yes, I did consider using Lua, but as you said, it doesn't have built-in support for UTF-8 either. CMake seemed like a better choice because it could be used by other CMake project as well.

justinmk commented on third-party/cmake/LibvtermCMakeLists.txt:
```diff
@@ -60,13 +68,24 @@ set_target_properties(harness PROPERTIES
+      COMMAND ${CMAKE_COMMAND} -E compare_files
+              ${header_path} ${perl_header_path}
+      DEPENDS ${header_path} ${perl_header_path})
+  endforeach()
```
How does one actually run these? Perhaps they should be dependencies on one of the main targets (`lint` or `test` ?) Otherwise this code is "dead", no one will know to run it :) I guess the `check` target above was never run either.

b-r-o-c-k replied:
I was running them using `cmake --build . --target test-{name}.inc`. I decided to use `add_custom_target` for consistency with the `check` target, but CMake also has the `add_test` command. To run those tests, the `ctest` tool would need to be used. If you want them to run during every build then adding them all as a dependencies of one main target is probably the best way of doing it.

justinmk commented:
This worked beautifully. I updated the wiki: https://github.com/neovim/neovim/wiki/Building-Neovim#windows--msvc
